Feature/issue 250 - Handle overlapping times with unique CRIDS

nikki-t opened the pull request:
Github Issue: #250

### Description

Within the same collection version there are multiple granules for the same cycle/pass (and continent) that have different CRIDs. This happens when there is a period of overlap between forward processing and reprocessing. The track ingest operations need to handle cases where this overlap occurs and prioritize a specific CRID when testing whether a granule was ingested into Hydrocron or not.

Cases handled

1) Reprocessing CRID granule arrives and replaces forward stream CRID
2) Reprocessing and froward stream CRIDs arrive at the same time, reprocessing CRID takes precedence and is ingested
3) Forward stream CRID arrives after reprocessing CRID, no action is taken as reprocessing CRID takes precendence
4) Forward stream CRID arrives during normal operations, forward stream is ingested

### Overview of work done

- Modified `query_cmr` to query parent collection (so that we don't have to wait for the reassociate tool to be run) in both OPS and SIT/UAT. Filter granules for either reaches, nodes, or prior lakes.
- Modified `query_hydrocron` to introduce logic to handle CRID cases above. Does introduce extra Hydrocron table queries in order to accomodate reprocessing CRID.

### Overview of verification done

- Added additional test for processing reprocessing CRID granule.
- Existing unit tests pass.

### Overview of integration done

Executed on cases handled locally on OPS CMR as no overlapping PGC0 and PIC0 granules are loaded into UAT. Did not execute `publish_cnm`.

Case 1) PGC0 replaces PIC0 where PIC0 is already ingested into Hydrocron SWOT table (3 other granules in track ingest table with to_ingest status)

```bash
2024-10-24T09:41:44,998 root INFO Querying CMR temporal range: 2024-01-21 22:50:00+00:00 to 2024-01-21 22:52:00+00:00.
2024-10-24T09:41:47,3 root INFO Located 2 granules in CMR.
2024-10-24T09:41:47,47 root INFO Located 0 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip.
2024-10-24T09:41:47,69 root INFO Located 1 items for forward stream granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PIC0_01.zip.
2024-10-24T09:41:47,90 root INFO Located 0 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip.
2024-10-24T09:41:47,90 root INFO Located 1 granules NOT in Hydrocron.
2024-10-24T09:41:47,90 root INFO Located 1 unique granules NOT in Hydrocron with PGC0 CRID.
2024-10-24T09:41:47,93 root INFO Located 4 granules with 'to_ingest' status.
2024-10-24T09:41:47,696 root INFO Located 1 granules that require ingestion.
2024-10-24T09:41:47,696 root INFO Located 4 granules that are already ingested.
2024-10-24T09:41:47,729 root INFO To Ingest: [{'granuleUR': 'SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip', 'revision_date': '2024-07-17T06:06:42.102Z', 'checksum': '7620183b46f90758241a5eac86c4efb3', 'expected_feature_count': -1, 'actual_feature_count': 0, 'status': 'to_ingest'}]
```

Case 2) CRIDs arrive at the same time, PGC0 gets ingested (3 other granules in track ingest table with to_ingest status)

```bash
2024-10-24T10:08:11,173 root INFO Querying CMR temporal range: 2024-01-21 22:50:00+00:00 to 2024-01-21 22:52:00+00:00.
2024-10-24T10:08:12,528 root INFO Located 2 granules in CMR.
2024-10-24T10:08:12,568 root INFO Located 0 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip.
2024-10-24T10:08:12,586 root INFO Located 0 items for forward stream granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PIC0_01.zip.
2024-10-24T10:08:12,602 root INFO Located 0 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip.
2024-10-24T10:08:12,602 root INFO Located 2 granules NOT in Hydrocron.
2024-10-24T10:08:12,602 root INFO Located 1 unique CRID granules NOT in Hydrocron.
2024-10-24T10:08:12,604 root INFO Located 3 granules with 'to_ingest' status.
2024-10-24T10:08:12,899 root INFO Located 1 granules that require ingestion.
2024-10-24T10:08:12,899 root INFO Located 3 granules that are already ingested.
2024-10-24T10:08:12,920 root INFO To Ingest: [{'granuleUR': 'SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip', 'revision_date': '2024-07-17T06:06:42.102Z', 'checksum': '7620183b46f90758241a5eac86c4efb3', 'expected_feature_count': -1, 'actual_feature_count': 0, 'status': 'to_ingest'}]
```

Case 3) Forward stream CRID arrives after reprocessing CRID, no action taken

```bash
2024-10-24T10:03:21,717 root INFO Querying CMR temporal range: 2024-01-21 22:50:00+00:00 to 2024-01-21 22:52:00+00:00.
2024-10-24T10:03:23,31 root INFO Located 2 granules in CMR.
2024-10-24T10:03:23,118 root INFO Located 1 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip.
2024-10-24T10:03:23,143 root INFO Located 1 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_009_499_AF_20240121T225107_20240121T225110_PGC0_01.zip.
2024-10-24T10:03:23,143 root INFO Located 0 granules NOT in Hydrocron.
2024-10-24T10:03:23,143 root INFO Located 0 unique CRID granules NOT in Hydrocron.
2024-10-24T10:03:23,146 root INFO Located 0 granules with 'to_ingest' status.
2024-10-24T10:03:23,146 root INFO Located 0 granules that require ingestion.
2024-10-24T10:03:23,146 root INFO Located 0 granules that are already ingested.
2024-10-24T10:03:23,160 root INFO To Ingest: []
```

Case 4) Normal forward stream operations (3 other granules in track ingest table with to_ingest status)

```bash
2024-10-24T10:11:54,330 root INFO Querying CMR temporal range: 2024-10-20 22:14:00+00:00 to 2024-10-20 22:20:00+00:00.
2024-10-24T10:11:56,410 root INFO Located 1 granules in CMR.
2024-10-24T10:11:56,461 root INFO Located 0 items for reprocessed granule SWOT_L2_HR_RiverSP_Reach_022_547_SI_20241020T221402_20241020T221722_PGC0_01.zip.
2024-10-24T10:11:56,485 root INFO Located 0 items for forward stream granule SWOT_L2_HR_RiverSP_Reach_022_547_SI_20241020T221402_20241020T221722_PIC2_01.zip.
2024-10-24T10:11:56,486 root INFO Located 1 granules NOT in Hydrocron.
2024-10-24T10:11:56,486 root INFO Located 1 unique CRID granules NOT in Hydrocron.
2024-10-24T10:11:56,489 root INFO Located 3 granules with 'to_ingest' status.
2024-10-24T10:11:56,806 root INFO Located 1 granules that require ingestion.
2024-10-24T10:11:56,806 root INFO Located 3 granules that are already ingested.
2024-10-24T10:11:56,830 root INFO To Ingest: [{'granuleUR': 'SWOT_L2_HR_RiverSP_Reach_022_547_SI_20241020T221402_20241020T221722_PIC2_01.zip', 'revision_date': '2024-10-24T07:31:10.653Z', 'checksum': '7d270e76f2cf28f17018c1de3d0b80d8', 'expected_feature_count': -1, 'actual_feature_count': 0, 'status': 'to_ingest'}]
```

Deployed feature branch to SIT and ran #238 tests on reaches, nodes, and prior lakes. Confirmed track ingest tables were updated as expected.

## PR checklist:

* [X] Linted
* [X] Updated unit tests
* [X] Updated changelog
* [X] Integration testing

_See [Pull Request Review Checklist](../CONTRIBUTING.md#reviewing) for pointers on reviewing this pull request_

torimcd reviewed (commented):
This looks good to me, just a couple clarifying questions: 
1. The 'current' CRID is just specified in the event when calling the lambda right? Should we keep track in a constant somewhere or just leave it up to the operator/document on the wiki to know which CRID is current? 

2. I think I see how the normal operations case is handled (fwd stream during fwd operations), but just confirming what happens when the fwd stream starts _before_ a reprocessed CRID exists, do we also just handle that in operator documentation (ie supply a 'dummy' reproc CRID to the event until one exists)? Will it fail if the reprocessed_crid parameter is an empty string? I might be overthinking this, but there will be a period of time with the new version d data when forward stream is turned on, but reproc is not, and we might not know what the reprocessed_crid is for that collection during that time.

nikki-t commented:
> 1. The 'current' CRID is just specified in the event when calling the lambda right? Should we keep track in a constant somewhere or just leave it up to the operator/document on the wiki to know which CRID is current?

That's correct whatever CRID we wish to prioritize is specified in the event that gets passed to the Lambda. I wanted to make it easy to change so that we would not have to create a new release to modify the CRID. If we feel this is something that won't change often and want to enforce a more consistent use of it, we can certainly turn it into a constant. Otherwise I think we can document it on the wiki. If we leave things as is, the specification of the CRID happens [here](https://github.com/podaac/hydrocron/blob/456099ce96add7930e4c60353a5800c4e29a0dab/terraform/hydrocron-eventbridge.tf#L17).

> 2. I think I see how the normal operations case is handled (fwd stream during fwd operations), but just confirming what happens when the fwd stream starts before a reprocessed CRID exists, do we also just handle that in operator documentation (ie supply a 'dummy' reproc CRID to the event until one exists)? 

I think I understand what you are saying here, I also spent a lot of time thinking about this 😄 but did not anticipate not having a `reprocessed_crid`. I think it's possible to specify a dummy or empty string for the `reprocessed_crid` value but I will need to spend some time testing out this case. I think we may want to add a `forward_crid` argument to track the forward stream CRID. I should have some time this week to test.

nikki-t commented:
I updated the operations to allow for the specification of any empty string for the `reprocessed_crid` in the Lambda event.

- I noticed an additional added case where we transition from one forward stream CRID to another. The track ingest operations assume that the forward stream collection will always be incremented and prioritizes the highest increment, for example: PIC0 to PIC2 where PIC2 would be selected by track ingest to ingest. See this code: https://github.com/podaac/hydrocron/blob/2882adf7d9651ddeff22086663270b8d261494ea/hydrocron/db/track_ingest.py#L248-L249
	- Track ingest always picks the granule with the CRID that comes lexographically last.
	- This way we don't have to track and specify multiple forward stream CRIDs, Hydrocron track ingest operations will just "work" for any forward stream CRID.

Updated CRID cases:

1. Reprocessing CRID granule arrives and replaces forward stream CRID
2. Reprocessing and froward stream CRIDs arrive at the same time, reprocessing CRID takes precedence and is ingested
3. Forward stream CRID arrives after reprocessing CRID, no action is taken as reprocessing CRID takes precendence
4. Forward stream CRID arrives during normal operations, forward stream is ingested
5. We don't know the reprocessed CRID, so we specify an empty string for the Lambda event parameter.
6. We transition from one forward stream CRID to another but we don't have to specify anything different for normal track ingest operations.

Tested all cases locally on reaches, nodes, and lakes. Also tested on reaches in services SIT.
